File: trabajo 16/radix sort.cs
using System;

class Program
{
    static void CountingSort(int[] arr, int exp)
    {
        int n = arr.Length;
        int[] output = new int[n];
        int[] count = new int[10];

        // Contar la cantidad de ocurrencias de cada dígito
        for (int i = 0; i < n; i++)
        {
            int index = (arr[i] / exp) % 10;
            count[index]++;
        }

        // Sumar los conteos acumulados
        for (int i = 1; i < 10; i++)
        {
            count[i] += count[i - 1];
        }

        // Construir el arreglo de salida
        for (int i = n - 1; i >= 0; i--)
        {
            int index = (arr[i] / exp) % 10;
            output[count[index] - 1] = arr[i];
            count[index]--;
        }

        // Copiar al arreglo original
        for (int i = 0; i < n; i++)
        {
            arr[i] = output[i];
        }
    }

    static void RadixSort(int[] arr)
    {
        int max = arr[0];
        foreach (int num in arr)
        {
            if (num > max)
                max = num;
        }

        for (int exp = 1; max / exp > 0; exp *= 10)
        {
            CountingSort(arr, exp);
        }
    }

    static void Main()
    {
        Random rand = new Random();
        int[] arr = new int[20];

        for (int i = 0; i < arr.Length; i++)
        {
            arr[i] = rand.Next(1, 10000);  
        }

        Console.WriteLine("Arreglo desordenado:");
        Console.WriteLine(string.Join(", ", arr));

        RadixSort(arr);

        Console.WriteLine("\nArreglo ordenado:");
        Console.WriteLine(string.Join(", ", arr));
    }
}

File: trabajo 16/radix sort.py
import random
import time

# -------------------- RADIX SORT --------------------
def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    # Contar la cantidad de ocurrencias de cada dígito
    for i in range(n):
        index = arr[i] // exp
        count[index % 10] += 1

    # Sumar los conteos acumulados
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Construir el arreglo de salida
    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1

    # Copiar los resultados al arreglo original
    for i in range(n):
        arr[i] = output[i]


def radix_sort(arr):
    max_num = max(arr)
    exp = 1
    while max_num // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

# Generar lista de números aleatorios
arr = [random.randint(1, 9999) for _ in range(20)]  

print("Arreglo desordenado:")
print(arr)

inicio = time.time()
radix_sort(arr)
fin = time.time()

print("\nArreglo ordenado:")
print(arr)

